src/main/java/shadows/apotheosis/adventure/affix/socket/SocketingRecipe.java

```diff
@@ -41,21 +43,22 @@ public boolean matches(Container pInv, Level pLevel) {
 	 * Returns an Item that is the result of this recipe
 	 */
 	@Override
-	public ItemStack assemble(Container pInv) {
-		ItemStack out = pInv.getItem(0).copy();
-		if (out.isEmpty()) return ItemStack.EMPTY;
-		out.setCount(1);
-		int sockets = SocketHelper.getSockets(out);
-		List<ItemStack> gems = SocketHelper.getGems(out, sockets);
-		for (int idx = 0; idx < gems.size(); idx++) {
-			Gem gem = GemItem.getGem(gems.get(idx));
-			if (gem == null) {
-				gems.set(idx, pInv.getItem(1));
-				break;
-			}
-		}
-		SocketHelper.setGems(out, gems);
-		return out;
+	public ItemStack assemble(Container inventory) {
+		ItemStack result = inventory.getItem(0).copy();
+		if (result.isEmpty()) return ItemStack.EMPTY;
+		result.setCount(1);
+		int sockets = SocketHelper.getSockets(result);
+		List<ItemStack> gems = SocketHelper.getGems(result, sockets);
+		ItemStack gemStack = inventory.getItem(1).copy();
+		var event = new ItemSocketingEvent(result, gemStack);
+		MinecraftForge.EVENT_BUS.post(event);
+		if (event.isCanceled()) return ItemStack.EMPTY;
+		result = event.getItemStack();
+		gemStack = event.getGemStack();
+		int socket = SocketHelper.getFirstEmptySocket(result);
+		gems.set(socket, gemStack);
+		SocketHelper.setGems(result, gems);
+		return result;
```

Comment:
I think this flow would make way more sense to construct the normal output and pass that (along with the inputs) to the event for modification, and then return the (potentially modified) event result.